presto-spi/src/main/java/com/facebook/presto/spi/block/DictionaryBlock.java

```diff
@@ -203,20 +204,20 @@ public long getSizeInBytes()
 
     private void calculateCompactSize()
     {
-        long sizeInBytes = 0;
         int uniqueIds = 0;
         boolean[] seen = new boolean[dictionary.getPositionCount()];
+        boolean[] used = new boolean[dictionary.getPositionCount()];
         for (int i = 0; i < positionCount; i++) {
             int position = getId(i);
             if (!seen[position]) {
                 if (!dictionary.isNull(position)) {
-                    sizeInBytes += dictionary.getRegionSizeInBytes(position, 1);
+                    used[position] = true;
```

Comment:
I believe a similar change is needed in `getLogicalSizeInBytes` and `getRegionSizeInBytes`.

Reply:
The `getLogicalSizeInBytes` case is a little more complicated. We need to know how many times each position is seen. So we either need to change `getPosisionsSizeInBytes` to take an int array, which would give worse performance when not needed, or have a separate function. I'll leave that out of this PR for now. It's arguably a separate issue from what this one is trying to fix.